parser/feature_metrics.go

```diff
@@ -0,0 +1,213 @@
+package parser
+
+import (
+	"go/ast"
+
+	"encr.dev/parser/encoding"
+	"encr.dev/parser/est"
+	"encr.dev/parser/internal/locations"
+	"encr.dev/parser/internal/walker"
+	"encr.dev/pkg/errinsrc/srcerrors"
+	"encr.dev/pkg/idents"
+	meta "encr.dev/proto/encore/parser/meta/v1"
+	schema "encr.dev/proto/encore/parser/schema/v1"
+)
+
+// metricConstructor describes a particular metric constructor function.
+type metricConstructor struct {
+	FuncName  string
+	HasLabels bool
+}
+
+var metricConstructors = []metricConstructor{
+	{"NewCounter", false},
+	{"NewCounterGroup", true},
+	{"NewGauge", false},
+	{"NewGaugeGroup", true},
+}
+
+func init() {
+	registerResource(
+		est.MetricResource,
+		"metric",
+		"https://encore.dev/docs/observability/metrics",
+		"metrics",
+		"encore.dev/metrics",
+	)
+
+	registerResourceReferenceParser(
+		est.MetricResource,
+		(*parser).parseMetricReference,
+	)
+
+	for _, constructor := range metricConstructors {
+		numTypeArgs := 1
+		if constructor.HasLabels {
+			numTypeArgs = 2
+		}
+
+		registerResourceCreationParser(
+			est.MetricResource,
+			constructor.FuncName, numTypeArgs,
+			createMetricParser(constructor),
+			locations.AllowedIn(locations.Variable).ButNotIn(locations.Function),
+		)
+	}
+}
+
+func createMetricParser(con metricConstructor) func(*parser, *est.File, *walker.Cursor, *ast.Ident, *ast.CallExpr) est.Resource {
+	return func(p *parser, file *est.File, cursor *walker.Cursor, ident *ast.Ident, callExpr *ast.CallExpr) est.Resource {
+		if len(callExpr.Args) != 2 {
+			p.errf(
+				callExpr.Pos(),
+				"metrics.%s requires two arguments: the metric name and the metric configuration",
+				con.FuncName,
+			)
+			return nil
+		}
+		var metricKind meta.Metric_MetricKind
+		switch con.FuncName {
+		case "NewCounter", "NewCounterGroup":
+			metricKind = meta.Metric_COUNTER
+		case "NewGauge", "NewGaugeGroup":
+			metricKind = meta.Metric_GAUGE
+		default:
+			panic("unknown metric constructor name")
+		}
```

Comment:
@eandre Should I add the metric kind to `con` rather than switching on the function name to get it?